presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.spi.VariableAllocator;
+import com.facebook.presto.spi.WarningCollector;
+import com.facebook.presto.spi.plan.CteConsumerNode;
+import com.facebook.presto.spi.plan.CteProducerNode;
+import com.facebook.presto.spi.plan.CteReferenceNode;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeIdAllocator;
+import com.facebook.presto.spi.plan.SequenceNode;
+import com.facebook.presto.sql.analyzer.CTEInformationCollector;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.ApplyNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+import com.google.common.graph.Traverser;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.SystemSessionProperties.isMaterializeAllCtes;
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class LogicalCteOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public LogicalCteOptimizer(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (!isMaterializeAllCtes(session)) {
+            return PlanOptimizerResult.optimizerResult(plan, false);
+        }
+        PlanNode rewrittenPlan = new CteEnumerator(session.getCteInformationCollector(), idAllocator, variableAllocator).transformPersistentCtes(plan);
+        return PlanOptimizerResult.optimizerResult(rewrittenPlan, !rewrittenPlan.equals(plan));
```

Comment:
nit: simple rewrittenPlan.equals(plan) may introduce false positive

Reply:
the transform function returns the original plannode if untriggered

Reply:
What @feilong-liu is referring to is that we had cases in other optimizers, where a part of the subplan was changed in an equivalent way (for example reordering of columns at intermediate nodes), and the equality check returned false even if the plans were equivalent. To avoid false negatives, we changed all PlanOptimizers to return PlanOptimizerResult which has a boolean flag whether the transform actually triggered or not.
Perhaps you can reuse that here to make the intention more obvious